dateparser/utils.py

```diff
@@ -86,3 +87,37 @@ def find_date_separator(format):
     m = re.search(r'(?:(?:%[dbBmaA])(\W))+', format)
     if m:
         return m.group(1)
+
+
+class Registry(object):
+
+    _global_dict = {}
+    _tag = 'from_registry'
+
+    def __new__(cls, *args, **kwargs):
+        key = cls.get_key(*args, **kwargs)
+
+        if key in cls._global_dict:
+            return cls._tag_instance(cls._global_dict[key])
+
+        return cls._global_dict.setdefault(
+            key,
+            object.__new__(cls, *args, **kwargs)
+        )
```

Comment:
If you want to, you can go with this approach, but I am still not convinced about its flow being clear :)

I mentioned class decorator earlier, so let me share this approach with you for you to compare. Word of warning though: I didn't spent much time on naming, so this is only proof of concept.
```python
def registry(cls):
    def choose(creator):
        def constructor(cls, *args, **kwargs):
            key = cls.get_key(*args, **kwargs)

            if not hasattr(cls, "__registry_dict"):
                setattr(cls, "__registry_dict", {})
            registry_dict = getattr(cls, "__registry_dict")

            if key not in registry_dict:
                registry_dict[key] = creator(cls, *args, **kwargs)
            return registry_dict[key]
        return staticmethod(constructor)

    def self_destruct(initializer):
        def destructor(self, *args, **kwargs):
            if getattr(self, "__initialized", False):
                return
            else:
                initializer(self, *args, **kwargs)
                setattr(self, "__initialized", True)
        return destructor

    if not (hasattr(cls, "get_key")
            and isinstance(cls.get_key, MethodType)
            and cls.get_key.__self__ is cls):
        raise NotImplementedError("Registry classes require to implement class method get_key")

    setattr(cls, '__new__', choose(cls.__new__))
    setattr(cls, '__init__', self_destruct(cls.__init__.__func__))
    return cls
```